alibi_detect/od/base.py

```diff
@@ -0,0 +1,32 @@
+from __future__ import annotations
+from typing import Union
+
+from typing_extensions import Protocol, runtime_checkable
+
+
+# Use Protocols instead of base classes for the backend associated objects. This is a bit more flexible and allows us to
+# avoid the torch/tensorflow imports in the base class.
+@runtime_checkable
+class TransformProtocol(Protocol):
+    """Protocol for transformer objects."""
+    def transform(self, x):
+        pass
+
+    def _transform(self, x):
+        pass
+
+
+@runtime_checkable
+class FittedTransformProtocol(TransformProtocol, Protocol):
+    """Protocol for fitted transformer objects."""
+    def fit(self, x_ref):
+        pass
+
+    def _fit(self, x_ref):
+        pass
+
+    def check_fitted(self):
+        pass
+
+
+transform_protocols = Union[TransformProtocol, FittedTransformProtocol]
```

Comment:
Nitpick: type aliases should be `PascalCase` and singular ideally, since `TransformProtocol` is already taken, perhaps `TransformProtocolType`? Maybe not such a big deal... @ascillitoe

Reply:
Agree `PascalCase` would be nice for consistency, similar to [this incoming change](https://github.com/SeldonIO/alibi-detect/pull/739/commits/f819e63981425c60252fc4be0b3cdea9e00fcc3a#diff-7e81bcb898fff1e2b3bff8d8499dc66e5adc0364535b668f41e8d79fe84d5dc6). We could also do `TransformProtocol = Union[UnfittedTransformProtocol, FittedTransformProtocol]`.